src/System.Numerics.Tensors/System/Numerics/ArrayUtilities.cs

```diff
@@ -15,5 +17,181 @@ public static T[] Empty<T>()
         {
             return EmptyArray<T>.Value;
         }
+
+        public static long GetProduct(int[] dimensions, int startIndex = 0)
+        {
+            if (dimensions.Length == 0)
+            {
+                return 0;
+            }
+
+            long product = 1;
+            for (int i = startIndex; i < dimensions?.Length; i++)
+            {
+                if (dimensions[i] < 0)
+                {
+                    throw new ArgumentOutOfRangeException($"{nameof(dimensions)}[{i}]");
+                }
+
+                // we use a long which should be much larger than is ever used here,
+                // but still force checked
+                checked
+                {
+                    product *= dimensions[i];
+                }
+            }
+
+            return product;
+        }
+
+        public static bool IsAscending(int[] values)
+        {
+            for (int i = 1; i < values.Length; i++)
+            {
+                if (values[i] < values[i - 1])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public static bool IsDescending(int[] values)
+        {
+            for (int i = 1; i < values.Length; i++)
+            {
+                if (values[i] > values[i - 1])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Gets the set of strides that can be used to calculate the offset of n-dimensions in a 1-dimensional layout
+        /// </summary>
+        /// <param name="dimensions"></param>
+        /// <returns></returns>
+        public static int[] GetStrides(IReadOnlyList<int> dimensions)
```

Comment:
int[] does not allocate when you iterate over it and there is no interface dispatch.